fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops

bdhirsh opened the pull request:
This should fix https://github.com/pytorch/pytorch/issues/56273. I tested these changes locally by making them directly on top of https://github.com/pytorch/pytorch/pull/56151, and running the xla tests (`xla/test/cpp/build/test_ptxla`).

**Current state:** For ops that are ported to structured, If external backends like XLA have implemented the `out` op but not the `functional` version, they will call into our code-generated `CompositeExplicitAutograd` kernel, which calls the structured operator's `meta()` function and then redispatches to the external backend's `out` function.

If XLA has registered their own kernel to the `functional` variant of the op, it'll override our codegen'd composite kernel. XLA has logic to code-generate "CPU fallback" kernels for "required" ops. It gets this information based off of `RegistrationDeclarations.yaml`. That info was technically incorrect up until this PR, since we were code-generating `inplace/functional` composite kernels for structured ops, but not updating `RegistrationDeclarations.yaml` with that information.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56307 fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops**
* #56306 fix meta() calls for non-storage tensors (i.e. xla)

Differential Revision: [D27883950](https://our.internmc.facebook.com/intern/diff/D27883950)

bdhirsh committed:
fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops

[ghstack-poisoned]

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -656,7 +661,7 @@ def compute_registration_declarations(f: NativeFunction) -> str:
         # TODO: What exactly is the semantics of the 'dispatch' field?
         'dispatch': str(f.dispatch.keys() != {DispatchKey.CompositeImplicitAutograd}),
-        'default': str(any(is_generic_dispatch_key(k) for k in f.dispatch))
+        'default': str(any(is_generic_dispatch_key(k) for k in f.dispatch) or has_autogenerated_composite_kernel(f))
```
I thought about maybe just putting the `CompositeImplicitAutograd` key directly in each NativeFunction's dispatch table in `model.py`, so we wouldn't need this logic here. But that has some downsides: e.g. there's some code in `register_dispatch_key.py` that would like to be able to differentiate "real" CompositeImplicitAutograd kernels from ones that we've generated, which would become harder to do.

ezyang replied:
With your upcoming rewrite this probably should be moot soon, so I don't mind this special case.

bdhirsh committed:
Update on "fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops"

This should fix https://github.com/pytorch/pytorch/issues/56273. I tested these changes locally by making them directly on top of https://github.com/pytorch/pytorch/pull/56151, and running the xla tests (`xla/test/cpp/build/test_ptxla`).

**Current state:** For ops that are ported to structured, If external backends like XLA have implemented the `out` op but not the `functional` version, they will call into our code-generated `CompositeExplicitAutograd` kernel, which calls the structured operator's `meta()` function and then redispatches to the external backend's `out` function.

If XLA has registered their own kernel to the `functional` variant of the op, it'll override our codegen'd composite kernel. XLA has logic to code-generate "CPU fallback" kernels for "required" ops. It gets this information based off of `RegistrationDeclarations.yaml`. That info was technically incorrect up until this PR, since we were code-generating `inplace/functional` composite kernels for structured ops, but not updating `RegistrationDeclarations.yaml` with that information.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops"

This should fix https://github.com/pytorch/pytorch/issues/56273. I tested these changes locally by making them directly on top of https://github.com/pytorch/pytorch/pull/56151, and running the xla tests (`xla/test/cpp/build/test_ptxla`).

**Current state:** For ops that are ported to structured, If external backends like XLA have implemented the `out` op but not the `functional` version, they will call into our code-generated `CompositeExplicitAutograd` kernel, which calls the structured operator's `meta()` function and then redispatches to the external backend's `out` function.

If XLA has registered their own kernel to the `functional` variant of the op, it'll override our codegen'd composite kernel. XLA has logic to code-generate "CPU fallback" kernels for "required" ops. It gets this information based off of `RegistrationDeclarations.yaml`. That info was technically incorrect up until this PR, since we were code-generating `inplace/functional` composite kernels for structured ops, but not updating `RegistrationDeclarations.yaml` with that information.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops"

This should fix https://github.com/pytorch/pytorch/issues/56273. I tested these changes locally by making them directly on top of https://github.com/pytorch/pytorch/pull/56151, and running the xla tests (`xla/test/cpp/build/test_ptxla`).

**Current state:** For ops that are ported to structured, If external backends like XLA have implemented the `out` op but not the `functional` version, they will call into our code-generated `CompositeExplicitAutograd` kernel, which calls the structured operator's `meta()` function and then redispatches to the external backend's `out` function.

If XLA has registered their own kernel to the `functional` variant of the op, it'll override our codegen'd composite kernel. XLA has logic to code-generate "CPU fallback" kernels for "required" ops. It gets this information based off of `RegistrationDeclarations.yaml`. That info was technically incorrect up until this PR, since we were code-generating `inplace/functional` composite kernels for structured ops, but not updating `RegistrationDeclarations.yaml` with that information.


Differential Revision: [D27883950](https://our.internmc.facebook.com/intern/diff/D27883950)

[ghstack-poisoned]
